Allow mp.start_processes to create processes in parallel

lijia19 opened the pull request:
Summary:
Background discussion in https://fb.workplace.com/groups/319878845696681/posts/1226087421742481

and pytorch issue filed https://github.com/pytorch/pytorch/issues/133010

one way to fix this problem is to add an option to parallel start processes on pytorch side.

Test Plan: Tested aps run in problem and things are in parallel now (next diff)

Differential Revision: D61301603


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

d4l3k commented on torch/multiprocessing/spawn.py:
```diff
@@ -205,12 +206,13 @@ def __init__(self, processes, error_files):
-    fn, args=(), nprocs=1, join=True, daemon=False, start_method="spawn"
+    fn, args=(), nprocs=1, join=True, daemon=False, start_method="spawn", start_func_in_parallel=False
 ):
+    assert not start_func_in_parallel or start_method != "fork", "fork is not supported for start_func_in_parallel"
```
Since this is a performance optimization can we instead warn and use the non-threadpool path?

lijia19 replied:
changed to fallback to sequential if not forkserver or env in next version

d4l3k commented on test/distributed/elastic/multiprocessing/api_test.py:
```diff
@@ -367,10 +375,11 @@ def test_function_with_tensor(self):
-                    args={},
-                    envs={},
+                    args={0: ()},
+                    envs={0: {}},
```
did this behavior change?

lijia19 replied:
Its more like a test issue exposed. The PContext class infers nprocs from args size https://fburl.com/630wzww1, so nprocs is 0 before. The non parallel version simply skip starting processes but the parallel version will throw max_workers need to be bigger than 0.

d4l3k reviewed (commented):
Generally is looking pretty good -- @kurman thoughts?

I'm not quite sure I love the name `start_func_in_parallel` since it's a bit long but it is pretty unambiguous in what it does

kurman commented:
@d4l3k @lijia19 are there any concerns for making this a default behavior? The only concern I have is debugability. If we do so, we could remove all the parameters and provide a env based kill-switch to use original implementation and remove that eventually if there are no reported issues.

d4l3k commented:
@kurman I was a bit nervous about making threading the default behavior since with forking it can cause some serious issues and also the potential that using threads would increase the current happy path launch times. I don't think this entrypoint path is used frequently in OSS (usually the command line path) so it may actually be fine to just do a hard cut over with a kill switch.

lijia19 commented:
/easycla

d4l3k reviewed (approved):
LGTM, this should be safe given that this is an opt in behavior

lijia19 committed:
Allow mp.start_processes to create processes in parallel (#133707)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/133707

This is to fix the pytorch issue filed https://github.com/pytorch/pytorch/issues/133010

one way to fix this problem is to enable parallel start processes in mp.start_processes.

What else in the diff:
refactored a test case api_test which was repeating a lot of tests due to the inheritance.
added unit test for forkserver when parallel start is on.

Test Plan:
Added unit test.
https://fburl.com/testinfra/evuie3eh, all pass and the parallel forkserver version is sometimes faster.

Reviewed By: d4l3k

Differential Revision: D61301603

ezyang reviewed (approved):
stamping to unblock, I did not review

ZainRizvi commented:
@pytorchbot revert -c nosignal -m "sorry but trunk has been consistently broken since this PR was merged. See: [GH job link](https://github.com/pytorch/pytorch/actions/runs/10529617600/job/29191757055) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/3546628a2a167ace6060737eeccf8ee8fd87ddc0)"

ZainRizvi commented:
cc: @PaliC fyi
